### PR TITLE
Paintsetup 2: entityPaintSetup

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Continues from #681 and #684
Implements entityPaintSetup very similar to its OpenRCT2 equivalent.

### CODE REVIEW

File: src/OpenLoco/Paint/Paint.h [hunk 1]
```diff
@@ -119,6 +119,10 @@ namespace OpenLoco::Paint
         [[nodiscard]] Ui::ViewportInteraction::InteractionArg getStationNameInteractionInfo(const uint32_t flags);
         [[nodiscard]] Ui::ViewportInteraction::InteractionArg getTownNameInteractionInfo(const uint32_t flags);
         Gfx::drawpixelinfo_t* getContext() { return _dpi; }
+        uint8_t rotation() { return currentRotation; }
```
Comment (marijnvdwerf) [inline src/OpenLoco/Paint/Paint.h:122]: getRotation

File: src/OpenLoco/Paint/PaintEntity.cpp [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+#include "../Things/ThingManager.h"
+#include "Paint.h"
+
+using namespace OpenLoco::Interop;
+using namespace OpenLoco::Ui::ViewportInteraction;
+
+namespace OpenLoco::Paint
+{
+    // 0x004B0CCE
+    void paintVehicleEntities(PaintSession& session, vehicle_base* base)
```
Comment (marijnvdwerf) [inline src/OpenLoco/Paint/PaintEntity.cpp:15]: paintVehicleEntity (singular)

File: src/OpenLoco/Paint/PaintEntity.cpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        regs.ax = base->x;
+        regs.cx = base->y;
+        regs.dl = base->z;
+        regs.ebx = (base->sprite_yaw + (session.rotation() << 4)) & 0x3F;
+        regs.esi = reinterpret_cast<int32_t>(base);
+        call(0x004B0CCE, regs);
+    }
+
+    // 0x00440325
+    void paintMiscEntities(PaintSession& session, MiscBase* base)
```
Comment (marijnvdwerf) [inline src/OpenLoco/Paint/PaintEntity.cpp:27]: Same as above

File: src/OpenLoco/Paint/PaintEntity.cpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+        if (loc.x >= 0x4000 || loc.y >= 0x4000)
+        {
+            return;
+        }
+
+        ThingManager::ThingTileList entities(loc);
+        for (auto* entity : entities)
+        {
+            auto left = dpi->x;
```
Comment (marijnvdwerf) [inline src/OpenLoco/Paint/PaintEntity.cpp:55]: Maybe add a todo to convert this to some `rect.contains(otherRect)`-like function

File: src/OpenLoco/Paint/PaintEntity.h [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+{
+    struct map_pos;
+}
+
+namespace OpenLoco::Paint
+{
+    struct PaintSession;
+    void paintEntities(PaintSession& session, const Map::map_pos& loc);
+    void paintEntities2(PaintSession& session, const Map::map_pos& loc);
+}
```
Comment (marijnvdwerf) [inline src/OpenLoco/Paint/PaintEntity.h:14]: Do we really need this header? PaintEntity isn't really a separate thing from Paint.cpp, right?

File: src/OpenLoco/Paint/Paint.h [hunk 6]
```diff
@@ -119,6 +119,10 @@ namespace OpenLoco::Paint
...
+        void setCurrentObject(void* object) { _currentObject = object; }
```
Comment (marijnvdwerf) [inline src/OpenLoco/Paint/Paint.h:123]: `Object` is a bit confusing, considering we use the same term for .DAT files. How about 'Item'? or 'Subject'?

File: src/OpenLoco/Paint/PaintEntity.cpp [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+namespace OpenLoco::Paint
+{
+    // 0x004B0CCE
+    void paintVehicleEntities(PaintSession& session, vehicle_base* base)
+    {
+        registers regs{};
+        regs.ax = base->x;
+        regs.cx = base->y;
+        regs.dl = base->z;
+        regs.ebx = (base->sprite_yaw + (session.rotation() << 4)) & 0x3F;
```
Comment (marijnvdwerf) [inline src/OpenLoco/Paint/PaintEntity.cpp:21]: Is the anding neccesary? Wil rotation() ever be higher than zero?

**Review by marijnvdwerf — CHANGES_REQUESTED** [review]

File: src/OpenLoco/Paint/PaintEntity.h (hunk 5 above)
Comment (duncanspumpkin) [reply src/OpenLoco/Paint/PaintEntity.h:14]: I've started to get a bit annoyed with lack of headers for some things. So i thought I'd start adding more of them. These functions are only used by paint and we don't need/want to expose them in the paint interface

File: src/OpenLoco/Paint/Paint.h (hunk 6 above)
Comment (duncanspumpkin) [reply src/OpenLoco/Paint/Paint.h:123]: Subject would probably make most sense

File: src/OpenLoco/Paint/PaintEntity.cpp (hunk 7 above)
Comment (duncanspumpkin) [reply src/OpenLoco/Paint/PaintEntity.cpp:21]: It needs to roll over correctly when the entity has been rotated. So yes it is required

Comment (marijnvdwerf) [reply src/OpenLoco/Paint/PaintEntity.cpp:21]: Maybe have getRotation() return (_rotation & 0b11)?

Comment (duncanspumpkin) [reply src/OpenLoco/Paint/PaintEntity.cpp:21]: But I'd need a set as well. The sprite yaw is a bit of a special rotation field.

File: src/OpenLoco/Paint/PaintEntity.cpp (hunk 4 above)
Comment (duncanspumpkin) [reply src/OpenLoco/Paint/PaintEntity.cpp:55]: Yes i was also thinking of similar.

**Review by marijnvdwerf — APPROVED** [review]

**Pull request merged**